### DISCUSSION TITLE
GWT compilation error

### DISCUSSION THREAD

**Discussion by Birdasaur** [webfx-project/webfx#6] [Q&A] [answered]
Looking for clues for debugging the GWT part of the toolchain.

- I cloned, built and successfully tested your ported SpaceFX demo in the browser. (Great stuff)
- Duplicated your project structures manually. (I understand you have tools/scripts that generate this)
- Replaced references and dependencies to SpaceFX towards my own module. (named NEO)
- Modified my original NEO application to have a services package, webfx.xml and javafx.application.Application file in a way that mimics the SpaceFX example.
-  I did not duplicate the webfx-demo-spacefx-application-javafx nor gluon modules as I didn't need them.  (Was that horribly incorrect?  Is there some magic in these modules for the GWT compilation?) 

During compilation of the gwt-app module I receive the following error:

`--- gwt-maven-plugin:1.0.0:compile (default-compile) @ webfx-neo-application-gwt ---
Compiling module webfx-neo-application-gwt
   Tracing compile failure path for type 'java.util.ServiceLoader'
      [ERROR] Errors in 'file:/C:/dev/NEO/webfx-neo/webfx-neo-application-gwt/target/classes/super/java/util/ServiceLoader.java'
         [ERROR] Line 27: edu.jhuapl.neo.App cannot be resolved to a type`

I've updated my version of the ServiceLoader.java file to point at my own main class obviously but the build process isn't finding the type.  I currently do not see the difference between the original spacefx projects and my NEO version. Can you provide a clue where I should check?

Thanks!

**Comment by Birdasaur** [comment, marked as answer]
I am using OpenJDK/JFX 14... does webfx require 15 for some reason?

**Reply by salmonb** [reply]
You're courageous testing without the build tool which typically will solve all of these kinds of issues.

WebFX requires OpenJDK13 or above (in theory OpenJDK11 should be enough but due to some bugs in the java compiler, it works only from OpenJDK13).

Here, GWT complains about not finding the source of your app (edu.jhuapl.neo.App).

First be sure Maven generates the sources artifact of your application module (with maven-source-plugin), but I guess this is the case if you duplicated the project.

But it's not enough, you need to update the GWT module (module.gwt.xml file) and add:

`   <source path="edu/jhuapl/neo"/>`

otherwise your source won't be visible by the GWT compiler. This GWT masking feature is to allow mixing GWT compatible and GWT incompatible code (ex: server). All paths not covered by at least one of these source entries will be masked regarding the GWT compilation. This is probably the explanation of your error.

**Reply by Birdasaur** [reply]
Thanks for the feedback. I truly understand that the proper way is to use your build tool. However without that tool I'm stuck waiting and let's just say I am highly motivated to get this working for several of our applications.

Thank you for pointing out the module.gwt.xml file. I swear to you I compared my project files against the SpaceFX projects multiple times looking for differences and somehow missed that file.  Your assessment was correct. I updated the the xml file and now I have a new huge list of errors. Each error seems to indicate that "no source code is available" for various classes I use. (both my own, Java and third party) 
I am guessing your build tool automatically builds the list of source path entries for each package for each class. I also suspect this is why it was so easy to convert port Gerrit's SpaceFX code... his original application uses no third party libs and the source was flattened to a single folder. 
Unless you intend to make the build tool available very soon I am going to attempt to manually enter the paths. (which will be a painful process but...)  I will keep you up to date.

Sean

**Reply by Birdasaur** [reply]
Just so I understand, what is this build tool currently comprised of?  Is it a collection of tools connected via script?

**Reply by salmonb** [reply]
Glad you made some progress.

Another important condition is to list all the required sources in the pom.xml dependencies of your GWT application module. I guess this is why you have the "no source code is available" error. So try to list all your dependencies sources (with "sources" as classifier for each dependency). You already have a list of dependencies in that Maven file, so you can see how it works with the samples, you just need to extend this list with the dependencies that are missing (and also update the module.gwt.xml like before so they become visible to the GWT compiler).

No the build tool is not a collection of tools but one big tool that automatically generates all files or sections of files for your application including these dependencies. Each time you see "Generated by WebFX" in a file, it's actually generated by this tool. But it's not scriptable at the moment, you need to alter its code to achieve what you want, that's one of the main reason why it's not ready.

**Reply by salmonb** [reply]
Hi @Birdasaur,

Just to let you know that we just published the tool we were talking about (I know it's a while ago we had this discussion, but the project has been paused for nearly a year, before restarting earlier this year).

It's a CLI tool, its [repository][cli-repo] is now public, and there is a guide to getting started in the [documentation][webfx-docs].

[cli-repo]: https://github.com/webfx-project/webfx-cli
[webfx-docs]: https://docs.webfx.dev/

**Comment by tretonio** [comment]
Congratulations on the great work, I'm having difficulties generating Web Html, can you help me?

What I didn't understand in webfx
1 - Do I need to compile outside of it and generate a jar?
2 - I followed the steps from https://docs.webfx.dev/#_getting_started
Creating your first WebFX app

rw-rw-r-- 1 jorge jorge 1372 Jul 19 11:37 pom.xml
drwxrwxr-x 3 jorge jorge 4096 Jul 19 11:37 webfx-example-application
drwxrwxr-x 3 jorge jorge 4096 Jul 19 11:37 webfx-example-application-gluon
drwxrwxr-x 3 jorge jorge 4096 Jul 19 11:37 webfx-example-application-gwt
drwxrwxr-x 3 jorge jorge 4096 Jul 19 11:37 webfx-example-application-openjfx
-rw-rw-r-- 1 jorge jorge 395 Jul 19 11:36 webfx.xml

the directory
webfx-example-application-gwt contains
drwxrwxr-x 4 jorge jorge 4096 Jul 19 11:37 src
-rw-rw-r-- 1 jorge jorge 229 Jul 19 11:37 am webfx.xml

webfx build --gwt
does not generate anything

How to proceed?, do you have to install the gwt plugin? and if so how to install?

**Reply by salmonb** [reply]
Hi @tretonio,

You're welcome.

`webfx build --gwt` should work without any other install.
Can you give more details like the output of that command line?

**Reply by tretonio** [reply]
Linux 5.15.0-72-generic #79~20.04.1-Ubuntu SMP Thu Apr 20 22:12:07 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
******
java version "17.0.7" 2023-04-18 LTS
Java(TM) SE Runtime Environment Oracle GraalVM 17.0.7+8.1 (build 17.0.7+8-LTS-jvmci-23.0-b12)
Java HotSpot(TM) 64-Bit Server VM Oracle GraalVM 17.0.7+8.1 (build 17.0.7+8-LTS-jvmci-23.0-b12, mixed mode, sharing)
****
Maven home: /home/jorge/apache-maven-3.8.6
Java version: 17.0.7, vendor: Oracle Corporation, runtime: /home/jorge/graalvm/graalvm-jdk-17.0.7+8.1
Default locale: pt_BR, platform encoding: UTF-8
OS name: "linux", version: "5.15.0-72-generic", arch: "amd64", family: "UNIX"

Into directory webfx-example
webfx build --gwt

display
~/webfx/webfx-cli$ mvn package -P gwt-compile 
[INFO] Scanning for projects...
[INFO] 
[INFO] ------------------------< dev.webfx:webfx-cli >-------------------------
[INFO] Building webfx-cli 0.1.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-enforcer-plugin:3.1.0:enforce (enforce-maven) @ webfx-cli ---
[INFO] 
[INFO] --- maven-resources-plugin:3.2.0:resources (default-resources) @ webfx-cli ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Using 'UTF-8' encoding to copy filtered properties files.
[INFO] Copying 25 resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.10.1:compile (default-compile) @ webfx-cli ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- maven-resources-plugin:3.2.0:testResources (default-testResources) @ webfx-cli ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Using 'UTF-8' encoding to copy filtered properties files.
[INFO] skip non existing resourceDirectory /home/jorge/webfx/webfx-cli/src/test/resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.10.1:testCompile (default-testCompile) @ webfx-cli ---
[INFO] No sources to compile
[INFO] 
[INFO] --- maven-surefire-plugin:3.0.0-M7:test (default-test) @ webfx-cli ---
[INFO] No tests to run.
[INFO] 
[INFO] --- maven-jar-plugin:3.2.2:jar (default-jar) @ webfx-cli ---
[INFO] 
[INFO] --- maven-source-plugin:3.2.1:jar-no-fork (attach-sources) @ webfx-cli ---
[INFO] 
[INFO] --- maven-shade-plugin:3.3.0:shade (fat) @ webfx-cli ---
[INFO] Including org.apache.commons:commons-compress:jar:1.21 in the shaded jar.
[INFO] Including info.picocli:picocli:jar:4.7.0 in the shaded jar.
[INFO] Including dev.webfx:webfx-lib-reusablestream:jar:0.1.0-SNAPSHOT in the shaded jar.
[INFO] Including dev.webfx:webfx-platform-meta:jar:0.1.0-SNAPSHOT in the shaded jar.
[INFO] Including dev.webfx:webfx-platform-console:jar:0.1.0-SNAPSHOT in the shaded jar.
[INFO] Including dev.webfx:webfx-platform-util:jar:0.1.0-SNAPSHOT in the shaded jar.
[INFO] Including dev.webfx:webfx-platform-resource:jar:0.1.0-SNAPSHOT in the shaded jar.
[WARNING] Discovered module-info.class. Shading will break its strong encapsulation.
[WARNING] Discovered module-info.class. Shading will break its strong encapsulation.
[WARNING] Discovered module-info.class. Shading will break its strong encapsulation.
[WARNING] Discovered module-info.class. Shading will break its strong encapsulation.
[WARNING] Discovered module-info.class. Shading will break its strong encapsulation.
[WARNING] Discovered module-info.class. Shading will break its strong encapsulation.
[WARNING] commons-compress-1.21.jar, picocli-4.7.0.jar, webfx-cli-0.1.0-SNAPSHOT.jar, webfx-lib-reusablestream-0.1.0-SNAPSHOT.jar, webfx-platform-console-0.1.0-SNAPSHOT.jar, webfx-platform-meta-0.1.0-SNAPSHOT.jar, webfx-platform-resource-0.1.0-SNAPSHOT.jar, webfx-platform-util-0.1.0-SNAPSHOT.jar define 1 overlapping resource: 
[WARNING]   - META-INF/MANIFEST.MF
[WARNING] maven-shade-plugin has detected that some class files are
[WARNING] present in two or more JARs. When this happens, only one
[WARNING] single version of the class is copied to the uber jar.
[WARNING] Usually this is not harmful and you can skip these warnings,
[WARNING] otherwise try to manually exclude artifacts based on
[WARNING] mvn dependency:tree -Ddetail=true and the above output.
[WARNING] See https://maven.apache.org/plugins/maven-shade-plugin/
[INFO] Attaching shaded artifact.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  5.622 s
[INFO] Finished at: 2023-07-19T17:40:00-03:00
[INFO] ------------------------------------------------------------------------
Call duration: 9090 ms

**Reply by salmonb** [reply]
It looks like there is something wrong in your webfx shortcut, because your webfx command is executed in the webfx-cli directory instead of the working directory (here webfx-example). The first line should be `.../webfx-example$ mvn package -P gwt-compile` instead of `~/webfx/webfx-cli$ mvn package -P gwt-compile`.

Have you typed the exact command from our docs to create your webfx alias?

**Reply by tretonio** [reply]
Yes, I followed https://docs.webfx.dev/
What should I do to fix the problem?

My webfx.sh
#!/usr/bin/env bash
java -jar "$(dirname "$0")/target/webfx-cli-0.1.0-SNAPSHOT-fat.jar"  "$@"

alias  into directory webfx-cli 
echo "alias webfx='sh \"$(cd "$(dirname "$1")" && pwd -P)/$(basename "$1")/webfx.sh\"'" >> ~/.bashrc
source ~/.bashrc

**Reply by salmonb** [reply]
Our docs didn't ask to create your own webfx.sh, but to create an alias to the already existing webfx.sh located in the webfx-cli repository using the provided command: `echo "alias webfx='sh \"$(cd "$(dirname "$1")" && pwd -P)/$(basename "$1")/webfx.sh\"'" >> ~/.bashrc` and it's important to run that command under the webfx-cli directory as explained in the docs. You can check that the alias has been added to your .bashrc by typing `tail ~/.bashrc` (you should see the alias declaration at the end of the file).  Then type `source ~/.bashrc` to apply it to your current session, or start a new terminal session, and the webfx command should work as expected.

Let me know if your problem is solved after you removed your own webfx.sh and followed the above instructions.

**Reply by tretonio** [reply]
tail ~/.bashrc
  if [ -f /usr/share/bash-completion/bash_completion ]; then
    . /usr/share/bash-completion/bash_completion
  elif [ -f /etc/bash_completion ]; then
    . /etc/bash_completion
  fi
fi

[ -s "/home/jorge/.web3j/source.sh" ] && source "/home/jorge/.web3j/source.sh"
alias webfx='sh "/home/jorge/webfx/webfx-cli-main//webfx.sh"'
alias webfx='sh "/home/jorge/webfx/webfx-cli//webfx.sh"'

**Reply by salmonb** [reply]
Ok, it looks good. Where is your webfx-example directory? Maybe the issue is that you put it under webfx-cli? Can you check it's outside of the webfx-cli repository?

**Reply by tretonio** [reply]
/home/jorge/webfx/webfx-cli/webfx-example

**Reply by salmonb** [reply]
That's the issue. Your webfx-example should be a separate repository, not inside webfx-cli

**Reply by salmonb** [reply]
That's the issue. Your webfx-example should be a separate repository, not inside webfx-cli

**Reply by tretonio** [reply]
I will test separately and then give a return.
Thank you for your attention and your time

**Reply by tretonio** [reply]
Yes it worked !!

Thanks
